### PR TITLE
Feature/244 247

### LINKED ISSUE #244: Load the 'single file' data
Client would like to be able to load Sofia's data (i.e. the data that presents in a single cv file as opposed to the pair of files 'key-data'). We already implemented this change in the BE and the system can already deal with this data type, what need to be changed (I guess) is the modal in the FE as now it only works with the 'key-data' pair files. @stephenlenzi could you please attach here an example of Sofia's data?

### LINKED ISSUE #247: download function
Request: Users should be able to download the normalized XYZ coordinates of their populations, and the normalized atlas populations. We need to add a download button somewhere I guess

This seems uncontroversial

### PR DESCRIPTION
**Pull request by afonsobspinto**
Closes #244 
Closes #247 

Implemented solution: 

244:
- Adds endpoint to upload single file:
`/api/experiments/{id}/upload-single-file/:`
- Renames previous upload pair files endpoint to:
`/api/experiments/{id}/upload-pair-files/:`
- Updates frontend create experiment dialog

247:
- Adds endpoint to retrieve active population data (zipped):
`/api/experiments/{id}/download_populations/{active_populations}/`
- Adds download button to UI

misc:
- Changes the destination of the output of cordmap to inside the population folder
- Adds migrations to move files from existent populations according to point above
- Updates cordmap

How to test this PR: 

244:

https://user-images.githubusercontent.com/19196034/231309380-22582748-6177-408b-a3b2-486012dd99b3.mp4

https://user-images.githubusercontent.com/19196034/231309316-00145ca5-e14c-4b83-92b7-3c6ba9cf9151.mp4

247:

https://user-images.githubusercontent.com/19196034/231309949-ace566b4-7404-492a-ab2a-fd099178880d.mp4

### Sanity checks:
- [x] The pull request is explicitly linked to the relevant issue(s)
- [x] The issue is well described: clearly states the problem and the general proposed solution(s)
- [x] From the issue and the current PR it is explicitly stated how to test the current change
- [x] The labels in the issue set the scope and the type of issue (bug, feature, etc.)
- [x] All the automated test checks are passing
- [ ] All the linked issues are included in one milestone
- [x] All the linked issues are assigned

### Breaking changes (select one):
- [x] The present changes do not change the preexisting api in any way
- [ ] This PR and the issue are tagged as a `breaking-change`

### Possible deployment updates issues (select one):
- [ ] There is no reason why deployments may break after the current update
- [x] This PR and the issue are tagged as `alert:deployment`

### Test coverage (select one):
- [ ] Tests for the relevant cases are included in this pr
- [x] The changes included in this pr are out of the current test coverage scope

### Documentation (select one):
- [ ] The documentation has been updated to match the current changes
- [x] The changes included in this PR are out of the current documentation scope

### CODE REVIEW

File: applications/portal/backend/workspaces/settings.py [hunk 1]
```diff
@@ -226,7 +226,9 @@
 POSITION_WITHIN_SUBSEGMENT = 0.5  # Position (in percentage) of the slice to pick from a subsegment
 
 # for postgres configure to use the database pool
-if DATABASES.get("default", None) and DATABASES.get("default").get("ENGINE", None) == "django.db.backends.postgresql":
-    DATABASES.get("default").update({
-        "ENGINE": "django_postgrespool2"
-    })
+# if DATABASES.get("default", None) and DATABASES.get("default").get("ENGINE", None) == "django.db.backends.postgresql":
```
[the remark this answers is not in the document]
Comment (zsinnema) [reply applications/portal/backend/workspaces/settings.py:229]: @afonsobspinto imo this is due to some other error. Ples\ae fix the "other" error

File: applications/portal/backend/api/views/rest/experiment.py [hunk 2]
```diff
@@ -124,16 +129,61 @@ def upload_files(self, request, **kwargs):
...
+            response_status = status.HTTP_201_CREATED if created else status.HTTP_200_OK
+        except InvalidInputError:
+            response_status = status.HTTP_400_BAD_REQUEST
+        return Response(status=response_status)
+
+    @action(detail=True, methods=['get'],  url_path='download_populations/(?P<active_populations>[^/.]+)')
+    def download_populations(self, request, pk=None, active_populations=None):
+        experiment = self.get_object()
+
+        active_populations = active_populations.split(',')
```
Comment (zsinnema) [inline applications/portal/backend/api/views/rest/experiment.py:169]: @afonsobspinto I don't like the string parameter and splitting it here into an array. Why not use an array as the parameter type

File: applications/portal/backend/api/migrations/0010_alter_collaborator_experiment_and_more.py [hunk 3]
```diff
@@ -0,0 +1,24 @@
+# Generated by Django 4.0.6 on 2022-10-06 15:24
```
Comment (zsinnema) [inline applications/portal/backend/api/migrations/0010_alter_collaborator_experiment_and_more.py:1]: @afonsobspinto why is this migration needed?

File: applications/portal/backend/api/helpers/generate_population_cells.py [hunk 4]
```diff
@@ -1,20 +1,39 @@
 import os.path
 import uuid
 
-
 from api.services.filesystem_service import move_file
 
 OUTPUT_DIRECTORY = "./out"
-CELLS_CSV = "cells_with_labels"
+PAIR_CELLS_CSV_SUFFIXE = "cells_with_labels"
```
Comment (zsinnema) [inline applications/portal/backend/api/helpers/generate_population_cells.py:7]: @afonsobspinto why is there a diff between a single and a pair file? Can't we come up with something that is auto detecting the "file" type  (single/pair)

File: applications/portal/backend/api/management/commands/initialize_single_file_upload.py [hunk 5]
```diff
@@ -0,0 +1,37 @@
...
+        parser.add_argument("experiment_id", type=str)
+        parser.add_argument("filepath", type=str)
+
+    def handle(self, *args, **options):
+        from cordmap.register_fiducial.io import get_population_labels_from_file
+
+        filepath = PosixPath(os.path.join(PERSISTENT_ROOT, options["filepath"]))
+        experiment_id = int(options["experiment_id"])
+
+        populations = get_population_labels_from_file(filepath)
```
Comment (zsinnema) [inline applications/portal/backend/api/management/commands/initialize_single_file_upload.py:25]: @afonsobspinto this doesn't look DRY to me....

File: applications/portal/backend/api/models/population.py [hunk 6]
```diff
@@ -106,11 +106,11 @@ def _has_file_changed(self):
             and self.cells.file.name != current.cells.file.name
         )
 
-    def generate_cells(self, data_filepath: str):
+    def generate_cells(self, data_filepath: str, is_fiducial: bool = False):
```
Comment (zsinnema) [inline applications/portal/backend/api/models/population.py:109]: @afonsobspinto `is_fiducial` shouldn't this be an attribute of the population? I mean now we can generate cells (same population) with `is_fiducial` True and False. Not sure if that's meant like that

File: applications/portal/backend/api/services/workflows_service.py [hunk 7]
```diff
@@ -65,17 +66,32 @@ def execute_generate_population_cells_workflow(tasks_tuple: Tuple):
     ).execute()
 
 
-def execute_upload_files_workflow(experiment_id: int, key_filepath: str, data_filepath: str):
+def execute_upload_pair_files_workflow(experiment_id: int, key_filepath: str, data_filepath: str):
     current_app = get_current_configuration()
     operations.PipelineOperation(
```
Comment (zsinnema) [inline applications/portal/backend/api/services/workflows_service.py:71]: @afonsobspinto this is also not so DRY

File: applications/portal/backend/requirements.txt [hunk 8]
```diff
@@ -2,7 +2,7 @@ uritemplate==4.1.1
 psycopg2-binary==2.9.3
 django-dry-rest-permissions==1.2.0
 bg-atlasapi==1.0.2
-
+numpy==1.23.1
```
Comment (zsinnema) [inline applications/portal/backend/requirements.txt:5]: @afonsobspinto why do we need numpy and pandas?

**Review by zsinnema — CHANGES_REQUESTED** [review]

File: applications/portal/backend/api/migrations/0010_alter_collaborator_experiment_and_more.py (hunk 3 above)
Comment (afonsobspinto) [reply applications/portal/backend/api/migrations/0010_alter_collaborator_experiment_and_more.py:1]: Honestly I'm not sure, that appeared from running makemigrations without any changes applied:

Uploading 2023-04-12 09-20-41.mp4…

File: applications/portal/backend/workspaces/settings.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (afonsobspinto) [reply applications/portal/backend/workspaces/settings.py:229]:
```
 File "<frozen importlib._bootstrap>", line 228, in _call_with_frames_removed
  File "/usr/local/lib/python3.9/site-packages/django/contrib/auth/models.py", line 3, in <module>
    from django.contrib.auth.base_user import AbstractBaseUser, BaseUserManager
  File "/usr/local/lib/python3.9/site-packages/django/contrib/auth/base_user.py", line 49, in <module>
    class AbstractBaseUser(models.Model):
  File "/usr/local/lib/python3.9/site-packages/django/db/models/base.py", line 141, in __new__
    new_class.add_to_class("_meta", Options(meta, app_label))
  File "/usr/local/lib/python3.9/site-packages/django/db/models/base.py", line 369, in add_to_class
    value.contribute_to_class(cls, name)
  File "/usr/local/lib/python3.9/site-packages/django/db/models/options.py", line 235, in contribute_to_class
    self.db_table, connection.ops.max_name_length()
  File "/usr/local/lib/python3.9/site-packages/django/utils/connection.py", line 15, in __getattr__
    return getattr(self._connections[self._alias], item)
  File "/usr/local/lib/python3.9/site-packages/django/utils/connection.py", line 62, in __getitem__
    conn = self.create_connection(alias)
  File "/usr/local/lib/python3.9/site-packages/django/db/utils.py", line 208, in create_connection
    backend = load_backend(db["ENGINE"])
  File "/usr/local/lib/python3.9/site-packages/django/db/utils.py", line 126, in load_backend
    raise ImproperlyConfigured(
Process SpawnProcess-2:
django.core.exceptions.ImproperlyConfigured: 'django_postgrespool2' isn't an available database backend or couldn't be imported. Check the above exception. To use one of the built-in backends, use 'django.db.backends.XXX', where XXX is one of:
    'mysql', 'oracle', 'postgresql', 'sqlite3'
Traceback (most recent call last):
  File "/usr/local/lib/python3.9/site-packages/django/db/utils.py", line 113, in load_backend
    return import_module("%s.base" % backend_name)
  File "/usr/local/lib/python3.9/importlib/__init__.py", line 127, in import_module
    return _bootstrap._gcd_import(name[level:], package, level)
  File "<frozen importlib._bootstrap>", line 1030, in _gcd_import
  File "<frozen importlib._bootstrap>", line 1007, in _find_and_load
  File "<frozen importlib._bootstrap>", line 986, in _find_and_load_unlocked
  File "<frozen importlib._bootstrap>", line 680, in _load_unlocked
  File "<frozen importlib._bootstrap_external>", line 850, in exec_module
  File "<frozen importlib._bootstrap>", line 228, in _call_with_frames_removed
  File "/usr/local/lib/python3.9/site-packages/django_postgrespool2/base.py", line 39, in <module>
    from sqlalchemy.pool import manage
ImportError: cannot import name 'manage' from 'sqlalchemy.pool' (/usr/local/lib/python3.9/site-packages/sqlalchemy/pool/__init__.py)
The above exception was the direct cause of the following exception:
Traceback (most recent call last):
  File "/usr/local/lib/python3.9/multiprocessing/process.py", line 315, in _bootstrap
    self.run()
  File "/usr/local/lib/python3.9/multiprocessing/process.py", line 108, in run
    self._target(*self._args, **self._kwargs)
  File "/usr/local/lib/python3.9/site-packages/uvicorn/subprocess.py", line 76, in subprocess_started
    target(sockets=sockets)
  File "/usr/local/lib/python3.9/site-packages/uvicorn/server.py", line 60, in run
    return asyncio.run(self.serve(sockets=sockets))
  File "/usr/local/lib/python3.9/asyncio/runners.py", line 44, in run
    return loop.run_until_complete(main)
  File "/usr/local/lib/python3.9/asyncio/base_events.py", line 642, in run_until_complete
    return future.result()
  File "/usr/local/lib/python3.9/site-packages/uvicorn/server.py", line 67, in serve
    config.load()
  File "/usr/local/lib/python3.9/site-packages/uvicorn/config.py", line 458, in load
    self.loaded_app = import_from_string(self.app)
  File "/usr/local/lib/python3.9/site-packages/uvicorn/importer.py", line 21, in import_from_string
    module = importlib.import_module(module_str)
  File "/usr/local/lib/python3.9/importlib/__init__.py", line 127, in import_module
    return _bootstrap._gcd_import(name[level:], package, level)
  File "<frozen importlib._bootstrap>", line 1030, in _gcd_import
  File "<frozen importlib._bootstrap>", line 1007, in _find_and_load
  File "<frozen importlib._bootstrap>", line 986, in _find_and_load_unlocked
  File "<frozen importlib._bootstrap>", line 680, in _load_unlocked
  File "<frozen importlib._bootstrap_external>", line 850, in exec_module
  File "<frozen importlib._bootstrap>", line 228, in _call_with_frames_removed
  File "/usr/src/app/./main.py", line 20, in <module>
    apps.populate(settings.INSTALLED_APPS)
  File "/usr/local/lib/python3.9/site-packages/django/apps/registry.py", line 116, in populate
    app_config.import_models()
  File "/usr/local/lib/python3.9/site-packages/django/apps/config.py", line 304, in import_models
    self.models_module = import_module(models_module_name)
  File "/usr/local/lib/python3.9/importlib/__init__.py", line 127, in import_module
    return _bootstrap._gcd_import(name[level:], package, level)
  File "<frozen importlib._bootstrap>", line 1030, in _gcd_import
  File "<frozen importlib._bootstrap>", line 1007, in _find_and_load
  File "<frozen importlib._bootstrap>", line 986, in _find_and_load_unlocked
  File "<frozen importlib._bootstrap>", line 680, in _load_unlocked
  File "<frozen importlib._bootstrap_external>", line 850, in exec_module
  File "<frozen importlib._bootstrap>", line 228, in _call_with_frames_removed
  File "/usr/local/lib/python3.9/site-packages/django/contrib/auth/models.py", line 3, in <module>
    from django.contrib.auth.base_user import AbstractBaseUser, BaseUserManager
  File "/usr/local/lib/python3.9/site-packages/django/contrib/auth/base_user.py", line 49, in <module>
    class AbstractBaseUser(models.Model):
  File "/usr/local/lib/python3.9/site-packages/django/db/models/base.py", line 141, in __new__
    new_class.add_to_class("_meta", Options(meta, app_label))
  File "/usr/local/lib/python3.9/site-packages/django/db/models/base.py", line 369, in add_to_class
    value.contribute_to_class(cls, name)
  File "/usr/local/lib/python3.9/site-packages/django/db/models/options.py", line 235, in contribute_to_class
    self.db_table, connection.ops.max_name_length()
  File "/usr/local/lib/python3.9/site-packages/django/utils/connection.py", line 15, in __getattr__
    return getattr(self._connections[self._alias], item)
  File "/usr/local/lib/python3.9/site-packages/django/utils/connection.py", line 62, in __getitem__
    conn = self.create_connection(alias)
  File "/usr/local/lib/python3.9/site-packages/django/db/utils.py", line 208, in create_connection
    backend = load_backend(db["ENGINE"])
  File "/usr/local/lib/python3.9/site-packages/django/db/utils.py", line 126, in load_backend
    raise ImproperlyConfigured(
django.core.exceptions.ImproperlyConfigured: 'django_postgrespool2' isn't an available database backend or couldn't be imported. Check the above exception. To use one of the built-in backends, use 'django.db.backends.XXX', where XXX is one of:
    'mysql', 'oracle', 'postgresql', 'sqlite3'
```

File: applications/portal/backend/requirements.txt (hunk 8 above)
Comment (afonsobspinto) [reply applications/portal/backend/requirements.txt:5]: I got into some trouble due to the package versions not being fixed, concretely:

- AttributeError: module 'numpy' has no attribute 'bool'.
- AttributeError: 'DataFrame' object has no attribute 'append'

The packages are needed by cordmap so I moved the requirements changes from here to there. cc @stephenlenzi

**Comment by afonsobspinto** [comment]
Opened issue https://github.com/MetaCell/salk-interactive-atlas/issues/252 to tackle additional changes requested in meeting. cc @zsinnema

**Pull request merged**